frontend/model/captureLogs.js

```diff
@@ -0,0 +1,218 @@
+import { CAPTURED_LOGS, SET_APP_LOGS_FILTER } from '~/frontend/utils/events.js'
+
+/*
+  - giConsole/count - total logs stored
+  - giConsole/lastEntry - latest log hash
+  - giConsole/limit - the limit of entries.
+  - giConsole/markers - an array of markers at every nth entry
+  - giConsole/markerNth - the nth index in which marker log
+*/
+
+// Configuration:
+const ENTRIES_LIMIT = 5000
+const ENTRIES_MARKER_NTH = 1000
+
+const getMarkers = () => JSON.parse(localStorage.getItem('giConsole/markers')) || []
+
+// these are initialized at captureLogsStart()
+let isCapturing = false
+let isConsoleOveridden = false
+let lastEntry = null
+let entriesCount = null
+let appLogsFilter = []
+
+function captureLog (type, ...msg) {
+  const logEntry = `${Date.now()}_${Math.floor(Math.random() * 100000)}` // uuid
+
+  // detect when is an Error and capture it properly
+  // ex: uncaught Vue errors or custom try/catch errors.
+  msg = msg.map((m) => m instanceof Error ? (m.stack || m.message) : m)
+
+  localStorage.setItem(`giConsole/${logEntry}`,
+    JSON.stringify({
+      type,
+      msg,
+      prev: lastEntry,
+      timestamp: new Date().toISOString()
+    })
+  )
+
+  lastEntry = logEntry
+  entriesCount++
+
+  localStorage.setItem('giConsole/lastEntry', logEntry)
+  localStorage.setItem('giConsole/count', entriesCount)
+
+  verifyEntryIsMarker()
+  verifyLogsSize()
+
+  // To avoid infinite loop because we log all selector calls, we run sbp calls
+  // here in a roundabout way by getting the function to which they're mapped.
+  // The reason this works is because the entire `sbp` domain is blacklisted
+  // from being logged in main.js.
+  sbp('sbp/selectors/fn', 'okTurtles.events/emit')(CAPTURED_LOGS, lastEntry)
+}
+
+function verifyEntryIsMarker () {
+  if (entriesCount % ENTRIES_MARKER_NTH === 0) {
+    const markers = getMarkers()
+    // Save entry as a marker to be later deleted when logs are too big.
+    markers.push(lastEntry)
+    localStorage.setItem('giConsole/markers', JSON.stringify(markers))
+  }
+}
+
+function verifyLogsSize () {
+  // Delete nth oldest logs recursively. This scenario can happen when the
+  // entries limit is changed. Example: There are 400 logs and the limit is 500.
+  // We change the limit to 100 and the marker nth to 25. 325 logs need to be deleted.
+  // We do it recursively in chunks of 25 until there's only 75 logs again.
+  while (entriesCount >= ENTRIES_LIMIT) {
+    const markers = getMarkers()
+    let toDelete = ENTRIES_MARKER_NTH
+    let oldestEntry = markers.shift() // the oldest marker
+    do {
+      const log = JSON.parse(localStorage.getItem(`giConsole/${oldestEntry}`)) || {}
+      localStorage.removeItem(`giConsole/${oldestEntry}`)
+      entriesCount--
+      toDelete--
+      oldestEntry = log.prev
+    } while (toDelete && oldestEntry)
+
+    localStorage.setItem('giConsole/markers', JSON.stringify(markers))
+    localStorage.setItem('giConsole/count', entriesCount)
```

Comment:
I'm currently trying to debug `entriesCount` isn't being updated correctly during my testing when I lower down the limit to 80.

This is a minor request, but you could you move these two lines, and line 72 outside of the while loop? This will make it a bit faster since it won't have to repeatedly touch `localStorage` and stringify. I don't think that'll fix the issue I'm running into but it's an optimization we might as well do while we're working on this PR.

Reply:
Changed the code to only touch LS after the while loop.